[Feature] add 'create_if_not_exists' for views

eterna2 opened the pull request:
### Feature:
Add flag `create_if_not_exists` for `create_view` and `create_materialized_view`.

### Background:
Hi, I am using `create_view` and `create_materialized_view` to generate views via cli. However, i will like the cli to not throw an error if the views already exists. This PR adds a kwargs `create_if_not_exists` for `create_view` and `create_materialized_view`, which will insert `CREATE VIEW {} IF NOT EXISTS` if the flag is set.

The default for the flag is `False` so that everything will be backward compatible.

eterna2 commented:
@kurtmckee Not sure if you are the right person. Can u help me have a look at this PR?

kurtmckee commented:
@eterna2, I don't feel qualified yet to review feature requests; my focus has been ticket triage and bug fixes. The project policy is to have unit tests associated with pull requests, so you'll need to add unit tests that demonstrate that this feature is working.

For example, it will be good to confirm that the function continues to raise an exception when the view already exists (the tests may already check this but you should confirm), and also test that the error is suppressed when `create_if_not_exists` is True.

@kvesteri, would you review this PR? I'm interested in your thoughts on the parameter name `"create_if_not_exists"`. The implementation seems similar to how `"materialized"` is supported, but the parameter name seems inaccurate to my eyes and I don't have alternate suggestions.

kvesteri commented on sqlalchemy_utils/view.py:
```diff
@@ -6,16 +6,17 @@
 
 class CreateView(DDLElement):
-    def __init__(self, name, selectable, materialized=False):
+    def __init__(self, name, selectable, materialized=False, create_if_not_exists=False):
```
Name the parameter `if_not_exists`

kvesteri commented on sqlalchemy_utils/view.py:
```diff
@@ -6,16 +6,17 @@
         self.selectable = selectable
         self.materialized = materialized
-
+        self.not_exists = create_if_not_exists
```
Use name `if_not_exists`

kvesteri commented on sqlalchemy_utils/view.py:
```diff
@@ -74,7 +75,8 @@ def create_materialized_view(
     indexes=None,
-    aliases=None
+    aliases=None,
+    create_if_not_exists=False,
```
Use name `if_not_exists`

eterna2 commented:
Hi, I had changed the kwargs to `if_not_exists` as requested. 

Also made several other changes:
- added kwargs `or_replace` for `CREATE [OR REPLACE|ALTER] [MATERIALIZED] VIEW`
- added compilation for `mysql`, `postgres`, `mssql`, and `snowflake` dialects
- updated flake8 rules specifically for view.py to ignore the long docstr and redefinitions for the dialect specific compilation.

I am actually using this for snowflake specifically.

eterna2 commented:
Sorry. Some unnecessarily format changes cuz i was using black to quickly do the autoformatting.

kurtmckee commented:
Thanks @eterna2! I've kicked off the CI runs. Please also add unit tests that demonstrate the new functionality.

eterna2 commented:
Hi, I have added the unit tests for the new params. I just wrote the test for the generic use case and postgres.

eterna2 commented:
Seem like crypto package is raising a deprecation warning. Do u want me to add a marker to the requirement for py3.6?

kurtmckee commented:
No, I can add that separately.

kvesteri commented on sqlalchemy_utils/view.py:
```diff
@@ -6,21 +6,149 @@
+
+
+@compiler.compiles(CreateView, 'postgresql')
+def compile_create_materialized_view_(element, compiler, **kw):
```
Just name all the dialect specific functions with the dialect prefix. That makes them more readable. So this function would then be `compile_create_materialized_view_postgresql`

eterna2 replied:
ok done.
